File: internal/lua_engine/openapi_audit.go
package lua_engine

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"sync"

	"github.com/londonmax12/hyperz/internal/httpclient"
	"github.com/londonmax12/hyperz/internal/scope"
)

// OpenAPIAudit discovers OpenAPI / Swagger documents at well-known
// per-host paths and audits the document itself for three classes of
// exposure that ship in the spec long before any request hits a real
// endpoint:
//
//  1. Embedded credentials. Specs are routinely committed alongside the
//     code that consumes them, so an inline `example:` value or a
//     `default:` baked into a server variable can promote a real
//     vendor key into the public surface. Every secrets_in_body pattern
//     fires here, with severity carried through from the catalogue.
//  2. Example authorization headers. Operations frequently illustrate
//     usage with `example: Bearer eyJ...` or `example: Basic dXNlcjpw`
//     values. The strings are often dummies, but JWTs and base64 basic-
//     auth blobs occasionally embed a real test account's credentials -
//     and even fully-synthetic examples leak schema and signing
//     algorithm information when they shouldn't.
//  3. Auth-less operations. When the spec declares an authentication
//     scheme via components.securitySchemes (OAS3) or
//     securityDefinitions (Swagger 2) but one or more operations carry
//     no security requirement (no operation-level `security:` AND no
//     global `security:` default), those operations are publicly
//     reachable. That may be intentional (a health probe) or a missing
//     guard; either way the operator should see the list.
//
// Per host: the check probes each well-known path at most once per
// scan, caches the parsed document plus the findings it produced, and
// re-emits those cached findings against subsequent pages on the same
// host with the new page URL stamped onto each finding so the report
// ties the issue to a page the user actually visited.
//
// Passive (LevelPassive) check.
type OpenAPIAudit struct {
	once  sync.Once
	mu    sync.Mutex
	cache map[string]openAPIAuditCacheEntry
}

// openAPIAuditCacheEntry memoizes the per-host probe result. A nil
// facts pointer represents a confirmed negative (well-known paths 404
// or the body wasn't a recognizable spec); a populated facts is re-used
// across every page on the host so the well-known endpoints are probed
// at most once per scan.
type openAPIAuditCacheEntry struct {
	facts *OpenAPIAuditFacts
}

// OpenAPIAuditFacts is the raw scan-facts shape the bridge returns to
// the Lua port. The fields are the probe metadata (URL the spec was
// served from, response status, raw body) the audit policy reads to
// run the secret / example-auth / authless-op passes. The audit lives
// in the .lua file; this struct is the algorithm input, not a finding.
type OpenAPIAuditFacts struct {
	ProbeURL string
	Status   int
	Body     []byte
}

// openAPISpecBodyCap bounds the spec body the check buffers. A real-
// world spec with hundreds of operations and inline schemas can reach
// a couple of megabytes pretty-printed; 4 MiB is a comfortable ceiling
// without letting a misbehaving edge pin a worker on a slow stream.
const openAPISpecBodyCap = 4 << 20

// openAPIWellKnownPaths is the curated list of conventional URLs at
// which frameworks expose OpenAPI / Swagger documents. Probed once per
// host when the check fires; 404 / non-spec responses are silently
// skipped. The list deliberately overlaps the crawler's discovery
// list - this check runs even on operators who disabled the crawler's
// API-discovery feature.
var openAPIWellKnownPaths = []string{
	"/openapi.json",
	"/openapi.yaml",
	"/openapi.yml",
	"/swagger.json",
	"/swagger.yaml",
	"/swagger.yml",
	"/swagger/v1/swagger.json",
	"/v2/api-docs",
	"/v3/api-docs",
	"/api-docs",
	"/api/swagger.json",
	"/api/openapi.json",
}

// DiscoverFacts returns the cached or freshly fetched OpenAPI spec
// facts for the host implied by pageURL, or nil when no well-known
// path served a recognisable spec. Per-host cache lifetime matches
// this receiver's lifetime (one *OpenAPIAudit per scan registration).
//
// The Lua port reads these facts and composes the finding catalog
// itself (title / severity / detail / CWE / OWASP / remediation /
// dedupe key / evidence); the algorithm input (HTTP fetch + version
// gate) stays in Go so per-host work happens at most once per scan
// regardless of which implementation runs at scan time.
func (c *OpenAPIAudit) DiscoverFacts(ctx context.Context, client *httpclient.Client, sc *scope.Scope, pageURL string) (*OpenAPIAuditFacts, error) {
	c.once.Do(func() {
		c.cache = map[string]openAPIAuditCacheEntry{}
	})

	u, err := url.Parse(pageURL)
	if err != nil || u.Scheme == "" || u.Host == "" {
		return nil, nil
	}
	if !sc.Allows(u) {
		return nil, nil
	}
	hostKey := strings.ToLower(u.Scheme + "://" + u.Host)

	c.mu.Lock()
	entry, cached := c.cache[hostKey]
	c.mu.Unlock()
	if cached {
		return entry.facts, nil
	}

	facts := c.probeHost(ctx, client, sc, u)
	c.mu.Lock()
	c.cache[hostKey] = openAPIAuditCacheEntry{facts: facts}
	c.mu.Unlock()
	return facts, nil
}

// probeHost walks the well-known spec paths on u's host and returns
// the facts for the first path whose body looks like a spec. Subsequent
// paths are not probed once a spec is found - mirror copies of the
// same document at multiple conventional URLs are expected and
// shouldn't produce duplicate findings.
func (c *OpenAPIAudit) probeHost(ctx context.Context, client *httpclient.Client, sc *scope.Scope, u *url.URL) *OpenAPIAuditFacts {
	base := u.Scheme + "://" + u.Host
	for _, path := range openAPIWellKnownPaths {
		probeURL := base + path
		probeU, err := url.Parse(probeURL)
		if err != nil {
			continue
		}
		if !sc.Allows(probeU) {
			continue
		}
		body, status, ok := c.fetchSpec(ctx, client, probeURL)
		if !ok {
			continue
		}
		if !looksLikeOpenAPIDoc(body) {
			continue
		}
		return &OpenAPIAuditFacts{
			ProbeURL: probeURL,
			Status:   status,
			Body:     body,
		}
	}
	return nil
}

// fetchSpec GETs probeURL and returns the response body when the
// response is a 200 the check can read. Returns ok=false on transport
// error, non-200, or empty body. The body is not parsed here so the
// caller can run regex passes that work on YAML and JSON alike.
func (c *OpenAPIAudit) fetchSpec(ctx context.Context, client *httpclient.Client, probeURL string) ([]byte, int, bool) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, probeURL, nil)
	if err != nil {
		return nil, 0, false
	}
	req.Header.Set("Accept", "application/json, application/yaml, */*")
	resp, err := client.Do(ctx, req)
	if err != nil {
		return nil, 0, false
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		_, _, _ = httpclient.ReadBodyCapped(resp, 1<<10)
		return nil, resp.StatusCode, false
	}
	body, _, err := httpclient.ReadBodyCapped(resp, openAPISpecBodyCap)
	if err != nil || len(body) == 0 {
		return nil, resp.StatusCode, false
	}
	return body, resp.StatusCode, true
}

// looksLikeOpenAPIDoc reports whether body carries the canonical
// version marker of an OpenAPI 3 or Swagger 2 document. The check
// must avoid auditing generic JSON 404 envelopes, README prose that
// mentions "OpenAPI", or YAML config that happens to carry an
// `openapi:` boolean key. Requiring the version key to be followed by
// a digit (`"openapi":"3.x"` / `openapi: 3.x` / `"swagger":"2.x"`)
// rejects all three false positives while matching every real spec.
func looksLikeOpenAPIDoc(body []byte) bool {
	lo := bytes.ToLower(bytes.TrimSpace(body))
	if len(lo) == 0 {
		return false
	}
	return openAPIVersionRE.Match(lo)
}

// openAPIVersionRE matches the version key of a spec in either format:
// JSON ("openapi": "3.x") and YAML (openapi: 3.x), with or without
// quotes around the version, requiring a leading digit so prose and
// boolean values do not match.
var openAPIVersionRE = regexp.MustCompile(`(?:"?openapi"?|"?swagger"?)\s*:\s*["']?\d`)



// titleAuthScheme normalizes a matched scheme to "Bearer" / "Basic"
// for display. strings.ToTitle / strings.Title both either uppercase
// the whole token or are deprecated; the audit only handles two
// schemes, so the explicit branch is the clearest option.
func titleAuthScheme(raw string) string {
	switch strings.ToLower(raw) {
	case "bearer":
		return "Bearer"
	case "basic":
		return "Basic"
	}
	return raw
}

// openAPIExampleHeaderRe matches `Bearer <token>` and `Basic <base64>`
// values plausibly shaped like a real Authorization header. The 20-
// char value floor is tight enough to skip placeholders like
// "Bearer TOKEN" / "Basic YOUR_KEY" while still catching JWTs and
// base64 blobs that resemble production credentials.
var openAPIExampleHeaderRe = regexp.MustCompile(`(?i)\b(Bearer|Basic)\s+([A-Za-z0-9._+/=-]{20,})`)

// openAPIExampleContextRe matches an OpenAPI key that explains why an
// Authorization-shaped value would appear in a documentation artifact.
// A bearer-shaped string that sits next to one of these keys is the
// thing the check wants to surface; one that does not is more likely
// to be unrelated text the operator already knows about. The optional
// trailing `"` / `'` lets the regex match the JSON shape `"example":`
// as well as the YAML shape `example:` without two separate patterns.
var openAPIExampleContextRe = regexp.MustCompile(`(?i)\b(examples?|default|value|x-example)["']?\s*[:=]`)



// openAPISecurityDoc is the subset of the spec the auth-less audit
// inspects: declared schemes (under either OAS3 or Swagger 2 keys),
// the global security default, and per-operation security. Unknown
// fields are ignored.
type openAPISecurityDoc struct {
	Security            []map[string][]string              `json:"security"`
	Components          openAPIComponents                  `json:"components"`
	SecurityDefinitions map[string]json.RawMessage         `json:"securityDefinitions"`
	Paths               map[string]openAPISecurityPathItem `json:"paths"`
}

type openAPIComponents struct {
	SecuritySchemes map[string]json.RawMessage `json:"securitySchemes"`
}

type openAPISecurityPathItem struct {
	Get     *openAPISecurityOp `json:"get"`
	Post    *openAPISecurityOp `json:"post"`
	Put     *openAPISecurityOp `json:"put"`
	Delete  *openAPISecurityOp `json:"delete"`
	Patch   *openAPISecurityOp `json:"patch"`
	Head    *openAPISecurityOp `json:"head"`
	Options *openAPISecurityOp `json:"options"`
}

// methodOp pairs a method label with the operation pointer so the
// caller can iterate without a branch per method. Returned in the
// canonical method order so the auth-less list is stable across runs.
type methodOp struct {
	method string
	op     *openAPISecurityOp
}

func (it openAPISecurityPathItem) methods() []methodOp {
	all := []methodOp{
		{"GET", it.Get},
		{"POST", it.Post},
		{"PUT", it.Put},
		{"DELETE", it.Delete},
		{"PATCH", it.Patch},
		{"HEAD", it.Head},
		{"OPTIONS", it.Options},
	}
	out := all[:0]
	for _, mo := range all {
		if mo.op == nil {
			continue
		}
		out = append(out, mo)
	}
	return out
}

// openAPISecurityOp captures the operation-level fields the audit
// reads. Security is a pointer to distinguish "not provided" (inherit
// global) from "explicitly empty" (intentional public op); Go's nil
// slice does not preserve that, so we keep the raw shape.
type openAPISecurityOp struct {
	Security *[]map[string][]string `json:"security"`
}

func (d openAPISecurityDoc) declaresSecurity() bool {
	return len(d.Components.SecuritySchemes) > 0 || len(d.SecurityDefinitions) > 0
}

// requirementIsAuthenticated returns true when at least one entry in
// req carries a non-empty scheme name. The empty list and a list
// whose only entry is `{}` both mean "no auth required"; an entry
// with any named scheme means the operation enforces auth.
func requirementIsAuthenticated(req []map[string][]string) bool {
	for _, entry := range req {
		for name := range entry {
			if strings.TrimSpace(name) != "" {
				return true
			}
		}
	}
	return false
}

File: internal/checks/openapi_audit.lua
-- openapi-audit: discovers OpenAPI / Swagger documents at well-known
-- per-host paths and audits the document itself for three classes of
-- exposure that ship in the spec long before any request hits a real
-- endpoint:
--
--   1. Embedded credentials. Every secrets_in_body pattern fires
--      against the spec body; specs are published alongside the code
--      that consumes them so any example/default value is a public
--      disclosure of every literal it carries.
--   2. Example authorization headers. Authorization-shaped values
--      sitting next to example/default/value blocks leak signing-algo
--      / claim-shape (JWTs) or username (Basic) even when fully
--      synthetic; accidental real credentials get flagged here too.
--   3. Auth-less operations. When the spec declares an auth scheme
--      via components.securitySchemes (OAS3) or securityDefinitions
--      (Swagger 2) but operations carry no security requirement and
--      inherit no global default, those operations are publicly
--      reachable.
--
-- Per-host caching of the probe + parsed body lives behind
-- ctx.openapi.discover (one fetch per host per scan).

local check = {
  name  = "openapi-audit",
  level = "passive",
  scope = "host",
  tier  = "passive",
}

local BODY_SNIPPET_CAP = 512

local function snippet_json(body)
  if body == nil or body == "" then return "" end
  local trimmed = string.gsub(body, "^%s+", "")
  trimmed = string.gsub(trimmed, "%s+$", "")
  if #trimmed > BODY_SNIPPET_CAP then
    return string.sub(trimmed, 1, BODY_SNIPPET_CAP)
  end
  return trimmed
end

local SEVERITY_RANK = { info = 0, low = 1, medium = 2, high = 3, critical = 4 }

local function build_evidence(ctx, facts)
  return ctx.evidence.build {
    method  = "GET",
    url     = facts.probe_url,
    status  = facts.status,
    snippet = snippet_json(facts.body),
  }
end

-- finding_embedded_credentials reuses ctx.body.find_secrets so this
-- check runs the same pattern catalogue and nearby-context filter as
-- secrets_in_body, then composes its own audit-side report (sort by
-- severity desc, max-severity, dedupe parts).
local function finding_embedded_credentials(ctx, facts)
  local hits = ctx.body.find_secrets(facts.body)
  if #hits == 0 then return nil end

  -- find_secrets returns hits pre-sorted (severity desc, id asc,
  -- redacted asc) so the first hit's label feeds the single-hit title.

  local max_sev = "info"
  local details = {}
  local id_parts = {}
  for _, h in ipairs(hits) do
    if SEVERITY_RANK[h.severity] > SEVERITY_RANK[max_sev] then
      max_sev = h.severity
    end
    local occ = ""
    if h.count > 1 then
      occ = string.format(" (%d occurrences)", h.count)
    end
    details[#details + 1] = string.format("%s [%s]: %s%s", h.label, h.severity, h.redacted, occ)
    id_parts[#id_parts + 1] = h.id .. ":" .. h.redacted
  end

  local title
  if #hits == 1 then
    title = "OpenAPI spec embeds a credential (" .. hits[1].label .. ")"
  else
    title = string.format("OpenAPI spec embeds %d distinct credentials", #hits)
  end

  local detail = string.format(
    "The OpenAPI / Swagger document at %s contains values matching known credential patterns. "
      .. "Specs are frequently published alongside the code that consumes them, so any credential value "
      .. "baked into an example or default ships to every reader of the document. "
      .. "Treat each entry as compromised the moment the spec was served.",
    facts.probe_url)

  local remediation = "Remove the embedded credential from the spec and rotate the value immediately - "
    .. "publication of a spec is a public disclosure of every literal value it carries. "
    .. "Audit access logs for the affected key during the exposure window. "
    .. "Replace any example or default that needs to demonstrate the shape of an authorized request with "
    .. "an obviously-fake placeholder (e.g. `xxxxxxxxxxxx`) and document elsewhere how a reader can obtain "
    .. "real credentials. "
    .. "For specs generated from source annotations, scrub the upstream annotations so the next "
    .. "regeneration does not reintroduce the leak."

  local parts = { "embedded-credentials" }
  for _, p in ipairs(id_parts) do parts[#parts + 1] = p end

  return {
    severity    = ctx.severity[max_sev],
    target      = facts.probe_url,
    url         = facts.probe_url,
    title       = title,
    detail      = detail,
    details     = details,
    cwe         = "CWE-200, CWE-798",
    owasp       = "A02:2021 Cryptographic Failures",
    remediation = remediation,
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = parts,
    },
  }
end

-- finding_example_auth_tokens consumes the raw regex hits from
-- ctx.openapi.scan_example_auth_matches and applies the audit policy
-- (dedup by scheme+redacted, sort by scheme asc / redacted asc).
local function finding_example_auth_tokens(ctx, facts)
  local raw = ctx.openapi.scan_example_auth_matches(facts.body)
  if #raw == 0 then return nil end

  local seen = {}
  local hits = {}
  for _, h in ipairs(raw) do
    local key = h.scheme .. ":" .. h.redacted
    if not seen[key] then
      seen[key] = true
      hits[#hits + 1] = { scheme = h.scheme, redacted = h.redacted }
    end
  end
  if #hits == 0 then return nil end

  table.sort(hits, function(a, b)
    if a.scheme ~= b.scheme then return a.scheme < b.scheme end
    return a.redacted < b.redacted
  end)

  local details = {}
  local id_parts = {}
  for _, h in ipairs(hits) do
    details[#details + 1] = string.format("%s example: %s", h.scheme, h.redacted)
    id_parts[#id_parts + 1] = h.scheme .. ":" .. h.redacted
  end

  local detail = string.format(
    "The OpenAPI / Swagger document at %s carries Authorization-header values inside example / default / "
      .. "value blocks. Even fully-synthetic examples leak the signing algorithm and claim shape (for JWTs) "
      .. "or the username portion before the colon (for Basic) to every reader of the spec; an example "
      .. "accidentally populated with a real test-account credential is publicly disclosed.",
    facts.probe_url)

  local remediation = "Replace example Authorization values with obviously-fake placeholders that still "
    .. "demonstrate the wire format (e.g. `Bearer <token>` or `Basic dXNlcjpwYXNz` containing only "
    .. "synthetic data). "
    .. "For JWT examples, generate a token with random keys at documentation-build time rather than "
    .. "copying a real one from a development environment. "
    .. "For Basic examples, never use a real account's username even if the password portion is fake - "
    .. "the username alone is enough to enumerate the directory."

  local parts = { "example-auth-tokens" }
  for _, p in ipairs(id_parts) do parts[#parts + 1] = p end

  return {
    severity    = ctx.severity.low,
    target      = facts.probe_url,
    url         = facts.probe_url,
    title       = "OpenAPI spec carries example Authorization tokens",
    detail      = detail,
    details     = details,
    cwe         = "CWE-200",
    owasp       = "A05:2021 Security Misconfiguration",
    remediation = remediation,
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = parts,
    },
  }
end

-- finding_authless_operations consults ctx.openapi.scan_security_facts
-- for the security-relevant subset of the spec and reports operations
-- that carry no security requirement when the spec otherwise declares
-- at least one security scheme. The bridge keeps encoding/json from
-- allocating any field outside that subset - a 4 MiB spec with
-- hundreds of schemas would otherwise force the Lua VM to materialize
-- the entire document just to read four fields here. JSON-only (Go
-- check is the same); a spec served as YAML skips this pass but still
-- gets the other two.
local function finding_authless_operations(ctx, facts)
  local sec = ctx.openapi.scan_security_facts(facts.body)
  if sec == nil then return nil end
  if not sec.declares_security then return nil end

  local global_required = sec.global_required

  local unauth = {}
  for _, op in ipairs(sec.operations) do
    -- An op with no `security:` key inherits global_required; an op
    -- that declares `security: []` overrides global to "no auth" -
    -- the canonical way to mark a deliberately-public op under a
    -- secured API.
    local authed
    if op.has_security then
      authed = op.required
    else
      authed = global_required
    end
    if not authed then
      unauth[#unauth + 1] = { method = op.method, path = op.path }
    end
  end

  if #unauth == 0 then return nil end

  table.sort(unauth, function(a, b)
    if a.path ~= b.path then return a.path < b.path end
    return a.method < b.method
  end)

  local details = {}
  local id_parts = {}
  for _, e in ipairs(unauth) do
    details[#details + 1] = e.method .. " " .. e.path
    id_parts[#id_parts + 1] = e.method .. " " .. e.path
  end

  local detail = string.format(
    "The OpenAPI / Swagger document at %s declares an authentication scheme "
      .. "(components.securitySchemes / securityDefinitions) but %d operation(s) carry no security "
      .. "requirement and inherit no global default. Those operations are reachable without credentials.",
    facts.probe_url, #unauth)

  local remediation = "Add an explicit `security:` block to every operation that should require "
    .. "authentication, or set a global `security:` default at the document root that those operations "
    .. "inherit. "
    .. "For operations that are genuinely meant to be public (a health probe, a login endpoint), document "
    .. "the intent with `security: []` so the reader can tell at a glance that the empty requirement is "
    .. "deliberate rather than an oversight. "
    .. "Audit the listed operations against the application's authentication middleware - the spec and "
    .. "the runtime can diverge, and either side of the divergence is a finding."

  local parts = { "authless-operations" }
  for _, p in ipairs(id_parts) do parts[#parts + 1] = p end

  return {
    severity    = ctx.severity.medium,
    target      = facts.probe_url,
    url         = facts.probe_url,
    title       = "OpenAPI spec declares auth schemes but exposes unauthenticated operations",
    detail      = detail,
    details     = details,
    cwe         = "CWE-306",
    owasp       = "A01:2021 Broken Access Control",
    remediation = remediation,
    evidence    = build_evidence(ctx, facts),
    dedupe_key  = ctx.dedupe.key {
      check  = check.name,
      scope  = "host",
      target = facts.probe_url,
      parts  = parts,
    },
  }
end

-- The audit builds each finding against the discovery probe URL (so
-- the dedupe key keys off the canonical resource); restamp Target /
-- URL to the current page so the report ties to a URL the operator
-- visited.
local function restamp_to_page(findings, page_url)
  for _, f in ipairs(findings) do
    f.target = page_url
    f.url    = page_url
  end
end

function check.run(ctx)
  local facts, err = ctx.openapi.discover(ctx.page.url)
  if err then
    ctx:report("openapi-audit: " .. err)
    return nil
  end
  if facts == nil then return nil end

  local findings = {}
  local f1 = finding_embedded_credentials(ctx, facts)
  if f1 then findings[#findings + 1] = f1 end
  local f2 = finding_example_auth_tokens(ctx, facts)
  if f2 then findings[#findings + 1] = f2 end
  local f3 = finding_authless_operations(ctx, facts)
  if f3 then findings[#findings + 1] = f3 end

  if #findings == 0 then return nil end
  restamp_to_page(findings, ctx.page.url)
  return findings
end

return check
